sycl-fusion/passes/kernel-fusion/SYCLSpecConstMaterializer.cpp

```diff
@@ -0,0 +1,313 @@
+      Context, StringRef{(const char *)SpecConstData, SpecConstDataSize});
+  auto *TupleMD = MDTuple::get(Context, {StringMD});
+  SpecConstMD->addOperand(TupleMD);
+
+  return PreservedAnalyses::all();
+}
+
+static Constant *getScalarConstant(const unsigned char **ValPtr, Type *Ty) {
+  if (Ty->isIntegerTy()) {
+    unsigned NumBytes = Ty->getIntegerBitWidth() / 8;
+    uint64_t IntValue = 0;
+    std::memcpy(&IntValue, *ValPtr, NumBytes);
+    *ValPtr = *ValPtr + NumBytes;
+    return ConstantInt::get(Ty, IntValue);
+  }
+  if (Ty->isDoubleTy()) {
+    double DoubleValue = *(reinterpret_cast<const double *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(double);
+    return ConstantFP::get(Ty, DoubleValue);
+  }
+  if (Ty->isFloatTy()) {
+    float FloatValue = *(reinterpret_cast<const float *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(float);
+    return ConstantFP::get(Ty, FloatValue);
+  }
+  if (Ty->isHalfTy()) {
+    uint16_t HalfValue = *(reinterpret_cast<const uint16_t *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(uint16_t);
+    return ConstantFP::get(Ty, HalfValue);
+  }
+
+  llvm_unreachable("Scalar type not found.");
+}
+
+static Constant *getConstantOfType(const unsigned char **ValPtr, Type *Ty) {
+  assert(ValPtr && Ty && "Invalid input.");
+  if (Ty->isIntegerTy() || Ty->isFloatTy() || Ty->isDoubleTy() ||
+      Ty->isHalfTy())
+    return getScalarConstant(ValPtr, Ty);
+  if (auto *ArrTy = dyn_cast<ArrayType>(Ty)) {
+    SmallVector<Constant *> Elems;
+    auto *ElemTy = ArrTy->getArrayElementType();
+    auto NumElems = ArrTy->getArrayNumElements();
+    for (uint64_t I = 0; I < NumElems; ++I)
+      Elems.push_back(getConstantOfType(ValPtr, ElemTy));
+    return ConstantArray::get(ArrayType::get(ElemTy, NumElems), Elems);
+  }
+  if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+    SmallVector<Constant *> StructElems;
+    for (auto *ElemTy : StructTy->elements())
+      StructElems.push_back(getConstantOfType(ValPtr, ElemTy));
+    return ConstantStruct::get(StructTy, StructElems);
+  }
+
+  llvm_unreachable("Unknown type in getConstantOfType.");
+}
+
+void SYCLSpecConstMaterializer::allocateSpecConstant(StringRef KernelName) {
+  for (auto I : llvm::enumerate(TypesAndOffsets)) {
+    auto *const Ty = I.value().first;
+    assert(Ty->isSized());
+    const auto Offset = I.value().second;
+    assert(Offset < SpecConstDataSize && "Out of bounds access.");
+    const unsigned char *ValPtr = &SpecConstData[Offset];
+    auto *Initializer = getConstantOfType(&ValPtr, I.value().first);
+    // AMD's CONSTANT_ADDRESS and Nvidia's ADDRESS_SPACE_CONST happen to have
+    // the same value.
+    constexpr unsigned AS = 4;
+    auto *SpecConstGlobal = new GlobalVariable(
+        *Mod, Ty, /*isConstant*/ true, GlobalValue::WeakODRLinkage, Initializer,
+        Twine("SpecConsBlob_" + KernelName + "_" + Twine(I.index())),
+        /*InsertBefore*/ nullptr, GlobalValue::NotThreadLocal, AS,
+        /*isExternallyInitialized*/ false);
```

Comment:
```suggestion
        Twine("SpecConsBlob_").concat(Twine(KernelName)).concat(Twine("_")).concat(Twine(I.index())),
        /*InsertBefore*/ nullptr, GlobalValue::NotThreadLocal, AS,
        /*isExternallyInitialized*/ false);
```
Something like that should work.

Reply:
I find the `+` operator more intuitive than the dot notation and in the end we end up with a `Twine`, so I'd prefer to leave as it is, if that's ok with you?

Reply:
Sure. Just so I understand it correctly, you're worried that because the type of `KernelName` is `StringRef`, `operator+` will be resolved to the one of `StringRef` as well?

Reply:
Well, TBH, I wasn't sure if `StringRef` had one. Now that I think about it, I think both cases should've worked. In fact, even:
```c++
"SpecConsBlob_" + KernelName + "_" + Twine(I.index()),
```
Should be equivalent as we have `Twine operator+(const char *, StringRef)`. Anyway, LGTM as is